gobblin-cluster/src/main/java/gobblin/cluster/GobblinClusterManager.java

```diff
@@ -149,15 +167,109 @@ public GobblinClusterManager(String clusterName, String applicationId, Config co
   }
 
   /**
+   * Handle leadership change.
+   * The applicationLauncher is only started on the leader.
+   * The leader cleans up existing jobs before starting the applicationLauncher.
+   * @param changeContext notification context
+   */
+  private void handleLeadershipChange(NotificationContext changeContext) {
+    if (this.helixManager.isLeader()) {
+      // can get multiple notifications on a leadership change, so only start the application launcher the first time
+      // the notification is received
+      LOGGER.info("Leader notification for {} isLeader {} HM.isLeader {}", this.helixManager.getInstanceName(),
+          isLeader, this.helixManager.isLeader());
+
+      if (!isLeader) {
+        LOGGER.info("New Helix Controller leader {}", this.helixManager.getInstanceName());
+
+        // Clean up existing jobs
+        HelixAdmin helixAdmin = this.helixManager.getClusterManagmentTool();
+
+        String clusterName = this.helixManager.getClusterName();
+        TaskDriver taskDriver = new TaskDriver(this.helixManager);
+
+        List<String> resourceNames = helixAdmin.getResourcesInCluster(clusterName);
+
+        for (String r : resourceNames) {
+          LOGGER.info("resource {} found", r);
+
+          // try to infer queue and job names from the resource name.
+          // the resource name is in <queue name>_<jobid> format where the job id may start with _job or _
+          int separatorIndex = r.indexOf("_job");
+
+          if (separatorIndex == -1) {
+            separatorIndex = r.indexOf('_');
```

Comment:
lastIndexOf() also won't be reliable since the job name is repeated. What we need is to compare against the queue name, but that is not available until we upgrade to the newer helix.